Avoid `RangeError` in `arrayBind` foreign implementation

pete-murphy opened the pull request:
**Description of the change**

Fixes #309. See the [related discussion on Discourse](https://discourse.purescript.org/t/stack-safe-cartesian-product/4841).

- ec814e4c4bbc7cf43387b9fc0657a3023c069e47 adds a failing test, reproducing the issue by passing 106,000 arguments to `Array.prototype.push.apply`
- a7159c7e56cc67cd17168c439c01e661bc32bada adds a case to use `flatMap` if available in the `arrayBind` implementation
- e9ad550295918e59ba5e4e2f670ba2e5724e068f simplifies the other case, and makes it stack-safe but gives up the `Array.prototype.push.apply` optimization

I don't know the motivation behind that optimization, so I am likely not aware of the tradeoffs. I tried [some benchmarks](https://observablehq.com/d/20091fbb067cb04f) in various JS run times, not seeing wildly-different results between existing solution (`#ff725c`), `flatMap` (`#4269d0`), and simplified version added in e9ad550295918e59ba5e4e2f670ba2e5724e068f (`#efb118`):

<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://github.com/user-attachments/assets/f961d1ff-2ffa-442e-97f0-0d8fac63bc6e">
  <img alt="Multiple line charts showing benchmark results for three alternative implementations of arrayBind in four different JS runtimes: Safari, Firefox, Node, and Chrome" src="https://github.com/user-attachments/assets/4f95846c-cd79-4209-8ec0-2984d62ea280">
</picture>

See also #311 

---

**Checklist:**

- [x] Added the change to the changelog's "Unreleased" section with a reference to this PR (e.g. "- Made a change (#0000)")
- [x] Linked any existing issues or proposals that this pull request should close
- [x] Updated or added relevant documentation
- [x] Added a test for the contribution (if applicable)

pete-murphy committed:
test(#309): Failing test

Demonstrating that the current implementation of `Array`'s `Bind` instance
causes `RangeError: Maximum call stack size exceeded` when the output of `f` in
`ma >>= f` is sufficiently large.

This is due to usage of `Function.prototype.apply`. From
[MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/apply#using_apply_and_built-in_functions):

> But beware: by using apply() (or the spread syntax) with an arbitrarily long
arguments list, you run the risk of exceeding the JavaScript engine's argument
length limit.

> The consequences of calling a function with too many arguments (that is, more
than tens of thousands of arguments) is unspecified and varies across engines.
(The JavaScriptCore engine has a hard-coded [argument limit of
65536](https://webkit.org/b/80797).)

Node v20.18.1 seems to have a higher limit around 106,000.

pete-murphy committed:
fix(#309): Use `flatMap` if supported by runtime

pete-murphy committed:
fix(#309): Use simple stack-safe fallback

pete-murphy committed:
chore(#309): Add to CHANGELOG.md

natefaubion commented on src/Control/Bind.js:
```diff
@@ -1,8 +1,17 @@
 export const arrayBind = function (arr) {
+  if (typeof Array.prototype.flatMap === "function") {
```
Do we need to check this on every bind invocation? I think this could probably be a static check.

pete-murphy replied:
Updated here https://github.com/purescript/purescript-prelude/pull/314/commits/f8783251d38b8d2902397f0f481ccfe8b760a779, but not sure if it's what you had in mind?

natefaubion commented on src/Control/Bind.js:
```diff
@@ -1,8 +1,17 @@
+      const xs = f(arr[i]);
+      for (let j = 0, m = xs.length; j < m; j++) {
+        result.push(xs[j]);
+      }
```
I don't think anything in prelude ffi uses const or for/let. We should probably continue to use `var` for consistency for now.

pete-murphy replied:
Changed to use `var` and pulled the length variable out to be more consistent with other FFI code in this repo.

natefaubion replied:
I don't think this made it in. I still see let/const.

pete-murphy replied:
Force-pushed, `const` only appears in top-level declaration

pete-murphy committed:
feat(#309): Address feedback from code review

Using static check to determine if `Array.prototype.flatMap` is available, and
use `var` instead of `let` in for loop to match existing code style.

pete-murphy commented on src/Control/Bind.js:
```diff
@@ -1,9 +1,21 @@
-    return result;
-  };
-};
+    : function (arr) {
+      return function (f) {
+        const result = [];
+        const l = arr.length;
+        for (var i = 0; i < l; i++) {
+          const xs = f(arr[i]);
+          const k = xs.length;
+          for (var j = 0; j < k; j++) {
+            result.push(xs[j]);
+          }
+        }
+        return result;
+      };
+    };
```
Maybe worth pulling this out to a standalone function, to export and test

natefaubion replied:
I don't think a legacy FFI implementation should be a public Prelude export. If we want to test it we should just duplicate it, for now, and maybe add a comment.
